src/App.tsx

```diff
@@ -1,26 +1,169 @@
+import { FilterCriteria } from './types/FilterCriteria';
+import * as todoServise from './api/todos';
+import { Header, Footer, TodoList, ErrorNotification } from './components';
+
+const filterTodos = (tasks: Todo[], filterCriteria: FilterCriteria) => {
+  return tasks.filter(task => {
+    const matchesStatus =
+      filterCriteria === FilterCriteria.All ||
+      (filterCriteria === FilterCriteria.Active && !task.completed) ||
+      (filterCriteria === FilterCriteria.Completed && task.completed);
+
+    return matchesStatus;
+  });
+};
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [loading, setLoading] = useState(false);
+  const [filter, setFilter] = useState<FilterCriteria>(FilterCriteria.All);
+  const [titleTodo, setTitleTodo] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+
+  useEffect(() => {
+    setLoading(true);
+
+    todoServise
+      .getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setErrorMessage('Unable to load todos');
+      })
+      .finally(() => {
+        setLoading(false);
+      });
+  }, []);
+
+  useEffect(() => {
+    if (errorMessage) {
+      const timer = setTimeout(() => {
+        setErrorMessage('');
+      }, 3000);
+
+      return () => clearTimeout(timer);
+    }
+  }, [errorMessage]);
+
+  const getFilteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [todos, filter]);
+
+  function addTodo({ title, userId, completed }: Todo) {
+    setLoading(true);
+
+    const newTempTodo = {
+      id: 0,
+      userId: todoServise.USER_ID,
+      title,
+      completed: false,
+    };
+    setTempTodo(newTempTodo);
+
+    todoServise
+      .createTodo({ title, userId, completed })
+      .then(newTodo => {
+        setTodos(currentTodos => [...currentTodos, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add todo');
+      })
+      .finally(() => {
+        setLoading(false);
+        setTitleTodo('');
+        setTempTodo(null);
+      });
   }
 
+  function updateTodo(updatedTodo: Todo) {
+    setLoadingTodoIds(ids => [...ids, updatedTodo.id]);
+
+    todoServise
+      .updateTodo(updatedTodo)
+      .then(todo => {
+        setTodos(currentTodos => {
+          const newTodo = [...currentTodos];
+          const index = newTodo.findIndex(post => post.id === updatedTodo.id);
+          newTodo.splice(index, 1, todo);
+
+          return newTodo;
+        });
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        setLoadingTodoIds(ids => ids.filter(id => id !== updatedTodo.id)); // Remove todo id from loading state
+        setTitleTodo('');
+      });
+  }
+
+  function deleteTodo(todoId: number) {
```

Comment:
It is important to follow the same way of creating functions everywhere, fix it everywhere
```suggestion
  const deleteTodo = (todoId: number) => {
```